### ISSUE TITLE
Retain cache when switching Flutter channels

### ISSUE THREAD

**Issue by ThinkDigitalSoftware** [flutter/flutter#44212]
## Use case
Sometimes I may need to switch from master to stable, which, in this case is unchanged, to test something out. In cases where I have no internet connection or slow or metered internet, this is either inconvenient or impossible. Further, it leaves flutter in an unusable state, since it seems that the previous data is erased. So now I'm stuck. 

## Proposal
Can we have a flag to keep the latest versions of the SDK and binaries and only have them removed if upgrading and only after the upgrade has successfully replaced the old one?

**Comment by christopherfujino** [comment]
When you change flutter channels from `master` -> `stable`, under the hood, the flutter tool is doing some variation of:

```
git fetch
git checkout stable --
```

Which will check out the tip of the remote `stable` branch.

@ThinkDigitalSoftware when you say "which, in this case is unchanged" do you mean that both `master` and `stable` branches are currently at the same commit? If only your branch changed, and not the `HEAD` revision, the cache should not be invalidated and the binaries should not be re-downloaded (https://github.com/flutter/flutter/blob/d68278e3eabd4ad775323be3e1ed27e6c882b9e7/bin/flutter#L111 for linux/mac checks the framework revision, https://github.com/flutter/flutter/blob/d68278e3eabd4ad775323be3e1ed27e6c882b9e7/bin/internal/update_dart_sdk.sh#L24 checks the engine revision).

If, however, by switching branches your `HEAD` revision changed (and also the engine version), we invalidate the cache and have to re-download binaries. We have to keep the engine binaries synced with the version the framework lists in its stamp file (`/path/to/flutter/bin/internal/engine.version`) because this is a pinned dependency.

Or am I misunderstanding your proposal?

**Comment by ThinkDigitalSoftware** [comment]
A quick scenario to illustrate my point, if I'm testing code that works in master but not stable, due to a breaking change, I may switch to master to test, and then switch RIGHT back to stable when I'm finished so I can keep working on what I was doing previously. This currently causes 2 the rebuilding/ downloading of the Flutter binaries. Once for master and then once again when switching back to stable. I propose that flutter should keep the binaries for stable intact and only redownload/rebuild them if the head changes from when it was last on stable. I would love to be able to switch between channels offline assuming I've downloaded the branch and binaries at least once. Similar to how git doesn't require internet unless fetching new commits, flutter should switch without requiring a download unless there are new commits.

**Comment by ThinkDigitalSoftware** [comment]
A second issue, which will probably need its own issue is that some binaries don't download until they're used. I found this out the hard way when updating a channel or switching to a new one and then leaving home. When I tried to debug on iOS, it needed to download some iOS binaries, which I had to then do over my mobile Hotspot, which fortunately I had.

**Comment by christopherfujino** [comment]
> I propose that flutter should keep the binaries for stable intact and only redownload/rebuild them if the head changes from when it was last on stable. I would love to be able to switch between channels offline assuming I've downloaded the branch and binaries at least once.

So if I understand, what you're asking for is whenever invalidating the cache, don't delete the old results, but keep them around namespaced by the revision, in case you ever return to that revision. Is this correct? You would still need to download the binaries if switching to a revision you've never been to, but then returning to a revision you've already used would require no further download.

The only reason we don't do this is that it's a lot of disk space (I'm currently reading over 800mb on my linux setup), and for example a user on the `master` channel who frequently upgrades would use an incredible amount of disk space. However, it might make sense as an opt-in only feature for users in your particular situation.

> A second issue, which will probably need its own github issue is that some binaries don't download until they're used. I found this out the hard way when updating a channel or switching to a new one and then leaving home. When I tried to debug on iOS, it needed to download some iOS binaries, which I had to then do over my mobile Hotspot, which fortunately I had.

If you know you'll be working offline, you can work around this by `flutter precache --all` while you have network access. When we distribute pre-built release archives on [the website](), they are generated by running `flutter precache --all`.

**Comment by christopherfujino** [comment]
Actually, another reason we don't keep old caches around is because the logic handling that is in shell scripts, which is difficult to test.

**Comment by ThinkDigitalSoftware** [comment]
> So if I understand, what you're asking for is whenever invalidating the cache, don't delete the old results, but keep them around namespaced by the revision, in case you ever return to that revision. Is this correct? You would still need to download the binaries if switching to a revision you've never been to, but then returning to a revision you've already used would require no further download.

Yes, however, this would be just for the latest revision, not each. So at max, 4 sets of binaries. This would correspond with how git keeps the source for the branch downloaded.  

> Actually, another reason we don't keep old caches around is because the logic handling that is in shell scripts, which is difficult to test.

Why isn't the logic as simple as, "don't invalidate cache unless upgrading an existing branch"?

**Comment by ThinkDigitalSoftware** [comment]
> If you know you'll be working offline, you can work around this by flutter precache --all while you have network access. When we distribute pre-built release archives on the website, they are generated by running flutter precache --all

Thanks for this tip!

**Comment by christopherfujino** [comment]
> Why isn't the logic as simple as, "don't invalidate cache unless upgrading an existing branch"?

You would need to restructure the cache to embed the branch name (essentially having a new cache for each channel). And how would it handle situations where the `HEAD` is detached from any branch (possible if for example you issue `flutter version 1.12.1`). Also, the shell script has no concept of when the tool is upgrading itself.

This would be possible to implement, but risky as it would add complexity to the shell script and it would be difficult to test.

**Comment by ThinkDigitalSoftware** [comment]
What if each binary was somehow linked to the revision, which is compared with the existing binary and revision and only updated if they don't match?

**Comment by christopherfujino** [comment]
> What if each binary was somehow linked to the revision, which is compared with the existing binary and revision and only updated if they don't match?

This would involve making hundreds of comparisons rather than two (for the framework revision and the engine revision), and this system would not solve the initial problem of not wanting to delete cache binaries from previous channels.

**Comment by ThinkDigitalSoftware** [comment]
Ah ok. I'm out of ideas. Hopefully in the future we'll have a solution for this. Thanks for your time!

**Comment by christopherfujino** [comment]
One way to workaround this, if you frequently switch back and forth between stable and master, would be to have two installs of flutter. You could then symlink separate references to the two entrypoints into a location on your path, something like:

```
ln -s $HOME/flutter-stable/bin/flutter $HOME/bin/flutter-stable
ln -s $HOME/flutter-master/bin/flutter $HOME/bin/flutter-master
```

And then they would each own their own cache.

**Comment by ThinkDigitalSoftware** [comment]
That's true, however, this wouldn't work well with the IDEs

**Comment by christopherfujino** [comment]
You can re-configure the location of your flutter repo in the VS Code plugin, I'm assuming you can do the same in IntelliJ.

**Comment by ThinkDigitalSoftware** [comment]
Oh yeah, that's true. Awesome idea, thanks!

On Jan 23, 2020 1:25 PM, Christopher Fujino <notifications@github.com> wrote:

You can re-configure the location of your flutter repo in the VS Code plugin, I'm assuming you can do the same in IntelliJ.

**Comment by amanv8060** [comment]
Dart Sdk still gets downloaded  . I think that it is completely unneccessory  as it is not touched very often during development
`Downloading Dart SDK from Flutter engine 20a953183580250aac2e15d36007664118bda5ab...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  256M  100  256M    0     0  2279k      0  0:01:55  0:01:55 --:--:-- 1238k
`

**Comment by christopherfujino** [comment]
@Aman9026 
> Dart Sdk still gets downloaded . I think that it is completely unneccessory as it is not touched very often during development

The Dart SDK is required to run the `flutter` CLI tool, the dart analyzer, to `pub get` pub packages, and to build Flutter apps.

**Comment by amanv8060** [comment]
> @Aman9026
> 
> > Dart Sdk still gets downloaded . I think that it is completely unneccessory as it is not touched very often during development
> 
> The Dart SDK is required to run the `flutter` CLI tool, the dart analyzer, to `pub get` pub packages, and to build Flutter apps.

I am suggesting cache this

**Comment by Aman9026** [comment]
> @Aman9026
> 
> > Dart Sdk still gets downloaded . I think that it is completely unneccessory as it is not touched very often during development
> 
> The Dart SDK is required to run the `flutter` CLI tool, the dart analyzer, to `pub get` pub packages, and to build Flutter apps.

Wrong mention :confused:

**Comment by christopherfujino** [comment]
> > @Aman9026
> > > Dart Sdk still gets downloaded . I think that it is completely unneccessory as it is not touched very often during development
> > 
> > 
> > The Dart SDK is required to run the `flutter` CLI tool, the dart analyzer, to `pub get` pub packages, and to build Flutter apps.
> 
> Wrong mention

oops sorry

**Comment by christopherfujino** [comment]
> I am suggesting cache this

The reasons we don't cache this, as outlined above, is concerns about disk usage and also the fact that the logic managing the flutter cache is implemented in BASH/windows batch, which are very difficult to test.

**Comment by christopherfujino** [comment]
> I am suggesting cache this
@amanv8060 

The reasons we don't cache this, as outlined above, is concerns about disk usage and also the fact that the logic managing the flutter cache is implemented in BASH/windows batch, which are very difficult to test.

**Comment by jonahwilliams** [comment]
Duplicate of https://github.com/flutter/flutter/issues/44118

**Comment by andrewkolos** [comment]
Reopening this as this appears to be no longer duplicate with #44118 (the issue changed since its creation).
Personally, I would like this feature for the type of scenarios[ @ThinkDigitalSoftware mentioned back in 2020](https://github.com/flutter/flutter/issues/44212#issuecomment-577405060). However, I do agree with @christopherfujino that this would be not easy to implement and maintain.

I also agree that such a feature should be opt-in. Caching is inherently prone to problems (e.g. invalidation), so making it more complicated would probably compound these problems. It would also use a large amount of disk space.

**Comment by Jayasubin** [comment]
> When you change flutter channels from `master` -> `stable`, under the hood, the flutter tool is doing some variation of:
> 
> ```
> git fetch
> git checkout stable --
> ```
> 
> Which will check out the tip of the remote `stable` branch.
> 
> @ThinkDigitalSoftware when you say "which, in this case is unchanged" do you mean that both `master` and `stable` branches are currently at the same commit? If only your branch changed, and not the `HEAD` revision, the cache should not be invalidated and the binaries should not be re-downloaded (
> 
> https://github.com/flutter/flutter/blob/d68278e3eabd4ad775323be3e1ed27e6c882b9e7/bin/flutter#L111
> 
> for linux/mac checks the framework revision,
> https://github.com/flutter/flutter/blob/d68278e3eabd4ad775323be3e1ed27e6c882b9e7/bin/internal/update_dart_sdk.sh#L24
> 
> checks the engine revision).
> If, however, by switching branches your `HEAD` revision changed (and also the engine version), we invalidate the cache and have to re-download binaries. We have to keep the engine binaries synced with the version the framework lists in its stamp file (`/path/to/flutter/bin/internal/engine.version`) because this is a pinned dependency.
> 
> Or am I misunderstanding your proposal?

What I understand by this is that when running `flutter channel master`, flutter checks out the latest `master` branch from GitHub. If I understood right, can we make it such that the `flutter channel` command only checks out the local `master` branch if available, so that it would be possible to get away with caching only 4 sets of binaries at most, like @ThinkDigitalSoftware mentioned.

**Comment by christopherfujino** [comment]
> What I understand by this is that when running flutter channel master, flutter checks out the latest master branch from GitHub. If I understood right, can we make it such that the flutter channel command only checks out the local master branch if available, so that it would be possible to get away with caching only 4 sets of binaries at most, like @ThinkDigitalSoftware mentioned.

What you're proposing is what the tool does: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/channel.dart#L177

That is, doing `flutter channel master` is roughly equivalent to `git checkout master`. The issue is that we only have a single Flutter cache.

It's clear it would be useful for users to optionally have multiple caches simulatenously. What's not so clear to me is what the optimal user interface for this would be.

**Comment by Jayasubin** [comment]
> It's clear it would be useful for users to optionally have multiple caches simulatenously. What's not so clear to me is what the optimal user interface for this would be.

How about
- `flutter channel master` does what it does right now and gets changes from remote
- `flutter channel master --keep-cache` uses the cached version
